### PR TITLE
replace VarExportFunctionDynamicReturnTypeExtension with conditional-type stub

### PR DESCRIPTION
**Pull request by staabm**
with this PR I want to try to replace a existing phpstan extension with stubs and conditional types.
I did so with some very basic examples, like `var_export` etc. as a proof-of-concept.

It does not work yet though.

### CODE REVIEW

File: stubs/core.stub [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (staabm) [inline stubs/core.stub:1]: this stub is taken 1:1 from psalm, to verify phpstan is kind of compatible with the psalm notations

File: stubs/core.stub [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * @param mixed $var
+ * @param bool $return
+ * @return ($return is true ? string : void)
```
Comment (staabm) [inline stubs/core.stub]: when running tests on phpstan-src@1.6.x it seems to make a difference whether the `@return` expression here is sorrounded by braces.. I think this is not expected

see

grafik

vs. (when running without braces)

grafik

(both results are not correct though ;))

Comment (staabm) [reply stubs/core.stub]: after reading the [1.6.x release blog post](https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types), I conclude the paranthesis are necessary.

File: stubs/core.stub [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+/**
+ * @param mixed $var
+ * @param bool $return
+ * @return ($return is true ? string : null)
```
Comment (staabm) [inline stubs/core.stub:6]: psalm supports a `@return ($return is true ? string : void)` type, while it seems in phpstan atm we need to explicitly use `null`

**Review by staabm — COMMENTED** [review]
I don't know why yet, but the result of the test actually look like

```
9) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/pr-1244.php:24" ('type', 'C:\dvl\Workspace\phpstan-src-...44.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\ConditionalTypeForParameter Object (...), 24)
Expected type string|true, got type ($return is true ? string : true) in C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/pr-1244.php on line 24.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'string|true'
+'($return is true ? string : true)'

C:\dvl\Workspace\phpstan-src-staabm\src\Testing\TypeInferenceTestCase.php:90
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\NodeScopeResolverTest.php:919

10) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/pr-1244.php:25" ('type', 'C:\dvl\Workspace\phpstan-src-...44.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\ConditionalTypeForParameter Object (...), 25)
Expected type true, got type ($return is true ? string : true) in C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/pr-1244.php on line 25.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'true'
+'($return is true ? string : true)'
```

I tried moving the test into a `NodeScopeResolverTest` but this did not make a difference compared to `LegacyNodeScopeResolverTest`.

it seems the return-type is not finally resolved.
not sure whether I am missing something in the stub-file, or whether I am hitting a bug within conditional-types, which make phpstan not resolve the `ConditionalTypeForParameter`.

any hint would be appreciated

**Comment by rvanvelzen** [comment]
I'll try to look into this tomorrow.

File: stubs/core.stub (hunk 3 above)
Comment (rvanvelzen) [reply stubs/core.stub:6]: This is because the function map specifies `var_export` to return `string|null` which isn't compatible with `string|void`.

**Comment by rvanvelzen** [comment]
So the big problem is that default values aren't exposed for native functions, so `$return` can't actually be resolved. I've created #1261 to remedy that, but that will only work for PHP>=8 because there are no default values available for native functions before that.

**Comment by staabm** [comment]
thank you for your support Richard <3.

**Comment by ondrejmirtes** [comment]
Finished and merged: https://github.com/phpstan/phpstan-src/commit/09635a14f4cc0219a93dc795575cb0dbb04479fa

Thank you!

**Pull request closed without merging**